crates/block-producer/src/batch_builder/mod.rs

```diff
@@ -67,280 +66,246 @@ impl BatchBuilder {
-            })
-            .id();
-
-        self.task_map.push((task_id, id));
+    #[instrument(target = COMPONENT, name = "batch_builder.get_batch_inputs", skip_all, err)]
+    async fn get_batch_inputs(
+        &self,
+        batch: SelectedBatch,
+    ) -> Result<(Vec<AuthenticatedTransaction>, BatchInputs), BuildBatchError> {
+        let block_references =
+            batch.transactions.iter().map(AuthenticatedTransaction::reference_block);
+        let unauthenticated_notes = batch
+            .transactions
+            .iter()
+            .flat_map(AuthenticatedTransaction::unauthenticated_notes);
+
+        self.store
+            .get_batch_inputs(block_references, unauthenticated_notes)
+            .await
+            .map_err(BuildBatchError::FetchBatchInputsFailed)
+            .map(|inputs| (batch.transactions, inputs))
+    }
 
-        Ok(())
+    #[instrument(target = COMPONENT, name = "batch_builder.propose_batch", skip_all, err)]
+    async fn propose_batch(
+        transactions: Vec<AuthenticatedTransaction>,
+        inputs: BatchInputs,
+    ) -> Result<ProposedBatch, BuildBatchError> {
+        let transactions =
+            transactions.iter().map(AuthenticatedTransaction::proven_transaction).collect();
+
+        ProposedBatch::new(
+            transactions,
+            inputs.batch_reference_block_header,
+            inputs.chain_mmr,
+            inputs.note_proofs,
+        )
+        .map_err(BuildBatchError::ProposeBatchError)
     }
 
-    #[instrument(target = COMPONENT, skip_all, err, fields(batch_id))]
-    async fn build_batch(
-        txs: Vec<AuthenticatedTransaction>,
-        batch_inputs: BatchInputs,
-        batch_prover: BatchProver,
+    #[instrument(target = COMPONENT, name = "batch_builder.prove_batch", skip_all, err)]
+    async fn prove_batch(
+        &self,
+        proposed_batch: ProposedBatch,
     ) -> Result<ProvenBatch, BuildBatchError> {
-        let num_txs = txs.len();
-
-        info!(target: COMPONENT, num_txs, "Building a transaction batch");
-        debug!(target: COMPONENT, txs = %format_array(txs.iter().map(|tx| tx.id().to_hex())));
+        Span::current().set_attribute("prover.kind", self.batch_prover.kind());
 
-        let BatchInputs {
-            batch_reference_block_header,
-            note_proofs,
-            chain_mmr,
-        } = batch_inputs;
-
-        let transactions = txs.iter().map(AuthenticatedTransaction::proven_transaction).collect();
+        match &self.batch_prover {
+            BatchProver::Remote(prover) => {
+                prover.prove(proposed_batch).await.map_err(BuildBatchError::RemoteProverError)
+            },
+            BatchProver::Local(prover) => tokio::task::spawn_blocking({
+                let prover = prover.clone();
+                move || prover.prove(proposed_batch).map_err(BuildBatchError::ProveBatchError)
+            })
+            .await
+            .map_err(BuildBatchError::JoinError)?,
+        }
```

Comment:
Not for this PR, but we should make sure the returned proof matches the required security level. This is not an issue for local provers (since we set the security level explicitly), but does need to be checked for remote provers (even if only to avoid misconfiguration errors).

Let's add an issue for this.

Reply:
#744